src/base/NekRSSeparateDomainProblem.C

```diff
@@ -0,0 +1,343 @@
+
+#include "NekRSSeparateDomainProblem.h"
+#include "Moose.h"
+#include "AuxiliarySystem.h"
+#include "TimeStepper.h"
+#include "NekInterface.h"
+#include "TimedPrint.h"
+#include "MooseUtils.h"
+#include "CardinalUtils.h"
+
+#include "nekrs.hpp"
+#include "nekInterface/nekInterfaceAdapter.hpp"
+
+registerMooseObject("CardinalApp", NekRSSeparateDomainProblem);
+
+bool NekRSSeparateDomainProblem::_first = true;
+
+InputParameters
+NekRSSeparateDomainProblem::validParams()
+{
+  InputParameters params = NekRSProblemBase::validParams();
+  params.addParam<bool>("minimize_transfers_in", false, "Whether to only synchronize nekRS "
+    "for the direction TO_EXTERNAL_APP on multiapp synchronization steps");
+  params.addParam<bool>("minimize_transfers_out", false, "Whether to only synchronize nekRS "
+    "for the direction FROM_EXTERNAL_APP on multiapp synchronization steps");
+  params.addParam<bool>("moving_mesh", false, "Moving mesh");
+  params.addParam<bool>("toNekRS_interface", false, "External app -> NekRS interface present?");
+  params.addParam<bool>("toNekRS_temperature", false, "External app -> NekRS temperature transfer?");
+  params.addParam<bool>("fromNekRS_interface", false, "NekRS -> external app interface present?");
+  params.addParam<bool>("fromNekRS_temperature", false, "NekRS -> external app temperature transfer?");
+  params.addRequiredParam<std::vector<FileName>>("ExternalApp_filename", "External app input file name");
+  MooseEnum app_type("SamApp THMApp");
+  params.addRequiredParam<MooseEnum>("ExternalApp_type", app_type, "External app type");
+  params.addParam<std::vector<int>>("fromNekRS_boundary", "Boundary ID through which nekRS will be coupled to external app");
+  params.addParam<std::vector<int>>("NekRS_inlet_boundary", "NekRS Boundary ID for Pressure drop calculation");
+
+  return params;
+}
+
+NekRSSeparateDomainProblem::NekRSSeparateDomainProblem(const InputParameters &params) : NekRSProblemBase(params),
+    _serialized_solution(NumericVector<Number>::build(_communicator).release()), 
+    _moving_mesh(getParam<bool>("moving_mesh")),
+    _minimize_transfers_in(getParam<bool>("minimize_transfers_in")),
+    _minimize_transfers_out(getParam<bool>("minimize_transfers_out")),
+    _toNekRS(getParam<bool>("toNekRS_interface")),
+    _toNekRS_temperature(getParam<bool>("toNekRS_temperature")),
+    _fromNekRS(getParam<bool>("fromNekRS_interface")),
+    _fromNekRS_temperature(getParam<bool>("fromNekRS_temperature")),
+    _ExternalApp_filename(getParam<std::vector<FileName>>("ExternalApp_filename")),
+    _ExternalApp_type(getParam<MooseEnum>("ExternalApp_type")),
+    _fromNekRS_boundary(isParamValid("fromNekRS_boundary") ? &getParam<std::vector<int>>("fromNekRS_boundary") : nullptr),
+    _NekRS_inlet_boundary(isParamValid("NekRS_inlet_boundary") ? &getParam<std::vector<int>>("NekRS_inlet_boundary") : nullptr)
+{
+
+  if (!_toNekRS && !_fromNekRS)
```

Comment:
Good error check! It's minor, but something I like to do when I write error messages is to put the names of input parameters in single quotes so that they pop out a bit more to the user. i.e. write `'toNekRS_interface'` instead of `toNekRS_interface`. Then it's really clear it's an input file parameter.